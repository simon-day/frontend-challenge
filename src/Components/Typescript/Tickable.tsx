import { FC } from "react";
import { Wrapper } from "./tickableStyles";
import usePrevious from "../../Hooks/usePrevious";
import { IoMdArrowDropup, IoMdArrowDropdown } from "react-icons/io";
import { useSpring, animated } from "react-spring";

// üê∏

interface Tickable {
  value: string;
  tickRate: number;
  lowest: number;
  highest: number;
  showDiff: boolean;
}

const TickableTS: FC<Tickable> = ({
  value,
  tickRate,
  lowest,
  highest,
  showDiff,
}) => {
  const prevNumber: string = usePrevious<string>(value);

  const animatedText = useSpring({
    opacity: 1,
    config: { duration: tickRate / 2.5 },
    from: { val: +prevNumber },
    to: { val: +value },
  });

  const direction: string =
    value > prevNumber ? "up" : value < prevNumber ? "down" : "";

  const difference: number = Number(value) - Number(prevNumber);

  return (
    <Wrapper>
      <div className={"highLowContainer"}>
        <p className={"highLowValue"}>{lowest}</p>
        <p className={"highLowTextHeader"}>LOW</p>
      </div>

      <div title={difference.toString()} className="mainTextContainer">
        <animated.div className={`${direction} mainText`}>
          {animatedText?.val.to((val: number) => Math.floor(val).toString())}
        </animated.div>
        {showDiff && !!difference && (
          <p className="differenceText">
            ({difference > 0 && "+"}
            {difference || ""})
          </p>
        )}

        {direction === "up" ? (
          <IoMdArrowDropup className="icon" color="#73e84f" />
        ) : direction === "down" ? (
          <IoMdArrowDropdown className="icon" color="#ff4747" />
        ) : null}
      </div>

      <div className={"highLowContainer"}>
        <p className={"highLowValue"}>{highest}</p>
        <p className={"highLowTextHeader"}>HIGH</p>
      </div>
    </Wrapper>
  );
};

export default TickableTS;
