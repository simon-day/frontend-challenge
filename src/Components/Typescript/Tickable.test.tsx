import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Tickable from "./Tickable";
import { Globals } from "react-spring";

Globals.assign({
  skipAnimation: true,
});

// NOTE:
// react-spring broke tests when swapped from h1 to animated.div for value,
// difficult to find working solution

const testComp = (
  <Tickable
    showDiff={false}
    tickRate={5000}
    highest={0}
    lowest={0}
    value={"500"}
  />
);

describe("Tickable", () => {
  it("renders correctly enzyme", () => {
    const wrapper = shallow(testComp);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders the animated div", () => {
    const wrapper = shallow(testComp);
    expect(wrapper.find(".mainText").length).toBe(1);
  });

  it("renders a high and low value and updates them", () => {
    const wrapper = shallow(testComp);
    const lowTextOriginal = wrapper.find(".highLowValue").first();
    const highTextOriginal = wrapper.find(".highLowValue").last();
    expect(lowTextOriginal.text()).toBe("0");
    expect(highTextOriginal.text()).toBe("0");
    wrapper.setProps({ highest: 1000, lowest: 35 });
    const lowTextNew = wrapper.find(".highLowValue").first();
    const highTextNew = wrapper.find(".highLowValue").last();
    expect(lowTextNew.text()).toBe("35");
    expect(highTextNew.text()).toBe("1000");
  });

  it("should not render up or down icon when there's no previous value to compare", () => {
    const wrapper = shallow(testComp);
    expect(wrapper.find("svg").length).toBe(0);
  });

  // TEST BROKEN FROM REACT-SPRING
  it("Should update the displayed value with val change and add class styling", () => {
    const wrapper = shallow(testComp);
    expect(wrapper.find("h1").text()).toBe("500");
    wrapper.setProps({ value: "520" });
    expect(wrapper.find("h1").text()).toBe("520");
    expect(wrapper.find("h1").hasClass("up"));

    // INCREASE VAL ONCE MORE
    wrapper.setProps({ value: "600" });
    expect(wrapper.find("h1").text()).toBe("600");
    expect(wrapper.find("h1").hasClass("up"));

    // DECREASE VAL TO NUMBER GREATER THAN ORIGINAL STATE
    wrapper.setProps({ value: "550" });
    expect(wrapper.find("h1").text()).toBe("550");
    expect(wrapper.find("h1").hasClass("down"));
  });
});
