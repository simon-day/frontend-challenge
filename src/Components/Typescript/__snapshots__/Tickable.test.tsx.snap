// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tickable renders correctly enzyme 1`] = `
<styled.div>
  <div
    className="highLowContainer"
  >
    <p
      className="highLowValue"
    >
      0
    </p>
    <p
      className="highLowTextHeader"
    >
      LOW
    </p>
  </div>
  <div
    className="mainTextContainer"
  >
    <Animated(div)
      className=" mainText"
    >
      <Component />
    </Animated(div)>
  </div>
  <div
    className="highLowContainer"
  >
    <p
      className="highLowValue"
    >
      0
    </p>
    <p
      className="highLowTextHeader"
    >
      HIGH
    </p>
  </div>
</styled.div>
`;
